### PR TITLE
Feat/channel

### PR DESCRIPTION
**Pull request by Darlington02**

### CODE REVIEW

File: src/channel/channel.cairo [hunk 1]
```diff
@@ -0,0 +1,599 @@
...
+        fn create_channel(ref self: ComponentState<TContractState>, community_id: u256) -> u256 {
+            let channel_id = self.channel_counter.read() + 1;
+            let channel_owner = get_caller_address();
+
+            let new_channel = ChannelDetails {
+                channel_id: channel_id,
+                community_id: community_id,
+                channel_owner: channel_owner,
+                channel_metadata_uri: "",
+                channel_nft_address: 1.try_into().unwrap(),
```
Comment (mubarak23) [inline src/channel/channel.cairo:132]: we should have a deploy channel NFT function that return the contract address and use it here.

File: src/channel/channel.cairo [hunk 2]
```diff
@@ -0,0 +1,599 @@
...
+        impl Community: CommunityComponent::HasComponent<TContractState>
+    > of IChannel<ComponentState<TContractState>> {
+        /// @notice creates a new channel
+        fn create_channel(ref self: ComponentState<TContractState>, community_id: u256) -> u256 {
+            let channel_id = self.channel_counter.read() + 1;
+            let channel_owner = get_caller_address();
+
+            let new_channel = ChannelDetails {
+                channel_id: channel_id,
+                community_id: community_id,
```
Comment (mubarak23) [inline src/channel/channel.cairo:129]: we need to check community exist

File: src/channel/channel.cairo [hunk 3]
```diff
@@ -0,0 +1,599 @@
...
+            let profile = get_caller_address();
+            let channel_member = self.channel_members.read((channel_id, profile));
+
+            // check that profile is a channel member
+            assert(self.is_channel_member(profile, channel_id), NOT_CHANNEL_MEMBER);
+
+            // check that channel has members
+            let total_members: u256 = channel.channel_total_members;
+            assert(total_members > 1, CHANNEL_HAS_NO_MEMBER);
+
```
Comment (mubarak23) [inline src/channel/channel.cairo:186]: after the checks, let move the remaining part of this function into an internal function, the same way we did for join_channel

File: src/channel/channel.cairo [hunk 4]
```diff
@@ -0,0 +1,599 @@
...
+        fn set_channel_censorship_status(
+            ref self: ComponentState<TContractState>, channel_id: u256, censorship_status: bool
+        ) {
+            let mut channel = self.channels.read(channel_id);
+
+            // check caller is owner
+            assert(channel.channel_owner == get_caller_address(), NOT_CHANNEL_OWNER);
+
+            // update storage
+            channel.channel_censorship = censorship_status;
```
Comment (mubarak23) [inline src/channel/channel.cairo:283]: let do this instead, 

```
let update_channel = ChannelDetails {channel.channel_censorship: censorship_status, ..channel};
 self.channels.write(channel_id, update_channel);
```
its a lot cleaner, this way we know we are retaining the remaining details.

File: src/channel/channel.cairo [hunk 5]
```diff
@@ -0,0 +1,599 @@
...
+                        }
+                    );
+                index += 1;
+            };
+        }
+
+        /// @notice internal function to deploy a channel nft
+        /// @param channel_id id of channel
+        /// @param salt for randomization
+        fn _deploy_channel_nft(
```
Comment (mubarak23) [inline src/channel/channel.cairo:548]: let call this function inside `create_channel`

File: src/channel/channel.cairo [hunk 6]
```diff
@@ -0,0 +1,599 @@
...
+            self: @ComponentState<TContractState>, channel_id: u256
+        ) -> ByteArray {
+            self.channels.read(channel_id).channel_metadata_uri
+        }
+
+        /// @notice checks if the profile is a member of the channel
+        /// @param profile the address of the profile
+        /// @param channel_id the id of the channel
+        /// @return bool the profile membership status
+        fn is_channel_member(
```
Comment (mubarak23) [inline src/channel/channel.cairo:331]: we should return a tuple of member status and ChannelMember i.e `(is_member, channel_member_detail)`, 

we will need the `channel_member_detail` particularly in test, when we are test each item in the channel member struct is set 

also checking if channel_token_id is minted or burned

File: src/channel/channel.cairo [hunk 7]
```diff
@@ -0,0 +1,599 @@
...
+    #[generate_trait]
+    pub impl InternalImpl<
+        TContractState,
+        +HasComponent<TContractState>,
+        +Drop<TContractState>,
+        impl Community: CommunityComponent::HasComponent<TContractState>
+    > of InternalTrait<TContractState> {
+        /// @notice initalizes channel component
+        /// @param channel_nft_classhash classhash of channel NFT
+        fn _initializer(ref self: ComponentState<TContractState>, channel_nft_classhash: felt252) {
```
Comment (mubarak23) [inline src/channel/channel.cairo:392]: we should we have an external function that call this

**Review by mubarak23 — CHANGES_REQUESTED** [review]

File: src/channel/channel.cairo [hunk 8]
```diff
@@ -0,0 +1,599 @@
...
+    #[embeddable_as(KarstChannel)]
+    impl ChannelImpl<
+        TContractState,
+        +HasComponent<TContractState>,
+        +Drop<TContractState>,
+        impl Community: CommunityComponent::HasComponent<TContractState>
+    > of IChannel<ComponentState<TContractState>> {
+        /// @notice creates a new channel
+        fn create_channel(ref self: ComponentState<TContractState>, community_id: u256) -> u256 {
+            let channel_id = self.channel_counter.read() + 1;
```
[the remark this answers is not in the document]
Comment (Darlington02) [reply src/channel/channel.cairo:124]: we do, check the `_initializer` internal function

File: src/channel/channel.cairo (hunk 1 above)
Comment (Darlington02) [reply src/channel/channel.cairo:132]: that makes sense, skipped it, will implement now:)

File: src/channel/channel.cairo [hunk 9]
```diff
@@ -0,0 +1,599 @@
...
+        /// @param profile address to add to the channel
+        /// @param channel_id id of the channel to be joined
+        fn _join_channel(
+            ref self: ComponentState<TContractState>, profile: ContractAddress, channel_id: u256
+        ) {
+            let mut channel: ChannelDetails = self.channels.read(channel_id);
+
+            // check that user is a member of the community this channel belongs to
+            let community_instance = get_dep_component!(@self, Community);
+            let membership_status = community_instance
```
[the remark this answers is not in the document]
Comment (Darlington02) [reply src/channel/channel.cairo:407]: oh yeah, I think you implemented that recently. atm it returns just bool. will fix after resolving conflicts with main

File: src/channel/channel.cairo (hunk 7 above)
Comment (Darlington02) [reply src/channel/channel.cairo:392]: nope. it will be called within the constructor of the inheriting the contract

**Pull request merged**